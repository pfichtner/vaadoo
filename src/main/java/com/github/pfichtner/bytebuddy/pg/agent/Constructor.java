package com.github.pfichtner.bytebuddy.pg.agent;

import static lombok.AccessLevel.PRIVATE;

import java.util.Set;

import javax.validation.ConstraintViolation;
import javax.validation.ConstraintViolationException;
import javax.validation.Validation;

import lombok.NoArgsConstructor;
import net.bytebuddy.asm.Advice;

@NoArgsConstructor(access = PRIVATE)
public final class Constructor {

	@Advice.OnMethodExit
	public static void exitConstructor(@Advice.This Object thisObj) {
		Set<ConstraintViolation<Object>> validationResult = Validation.buildDefaultValidatorFactory().getValidator()
				.validate(thisObj);
		if (!validationResult.isEmpty()) {
			throw new ConstraintViolationException(validationResult);
		}
	}

}