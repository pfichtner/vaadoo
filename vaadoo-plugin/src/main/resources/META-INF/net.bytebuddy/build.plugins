com.github.pfichtner.vaadoo.AddJsr380ValidationPlugin
