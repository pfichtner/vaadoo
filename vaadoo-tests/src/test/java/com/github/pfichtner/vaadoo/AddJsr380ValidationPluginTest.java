package com.github.pfichtner.vaadoo;

import static java.lang.ClassLoader.getSystemClassLoader;
import static java.util.Collections.singletonMap;
import static org.approvaltests.Approvals.verify;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

import java.io.IOException;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;
import java.util.stream.Stream;

import org.jetbrains.java.decompiler.main.decompiler.ConsoleDecompiler;
import org.jetbrains.java.decompiler.main.extern.IFernflowerPreferences;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import com.example.SomeClass;
import com.example.SomeLombokClass;
import com.example.SomeRecord;

import net.bytebuddy.ByteBuddy;
import net.bytebuddy.build.Plugin;
import net.bytebuddy.description.type.TypeDescription;
import net.bytebuddy.dynamic.loading.ByteArrayClassLoader;

class AddJsr380ValidationPluginTest {

	@Test
	void testGeneratedBytecode(@TempDir Path tempDir) throws Exception {
		verify(decompile(tempDir, SomeClass.class));
	}

	@Test
	void testGeneratedBytecodeOnLombokClass(@TempDir Path tempDir) throws Exception {
		verify(decompile(tempDir, SomeLombokClass.class));
	}

	@Test
	void testGeneratedBytecodeOnRecord() throws Exception {
		try (AddJsr380ValidationPlugin sut = new AddJsr380ValidationPlugin()) {
			var constructor = firstConstructor(transform(sut, SomeRecord.class));
			assertThatExceptionOfType(InvocationTargetException.class).isThrownBy(() -> {
				constructor.newInstance(null, sut, null, null, null, null, null, null, null, "", false, false, null,
						null, 0, Long.valueOf(42), Short.valueOf((short) 42));
			}).satisfies(e -> assertThat(e.getCause()).isInstanceOf(NullPointerException.class)
					.hasMessageContaining("someNotEmptyCharSequence must not be null"));
		}
	}

	private String decompile(Path tempDir, Class<?> clazz) throws URISyntaxException, IOException {
		Path sourcePath = Path.of(getClass().getResource("/" + clazz.getName().replace('.', '/') + ".class").toURI());
		Files.copy(sourcePath, tempDir.resolve(clazz.getSimpleName() + ".class"));
		return decompileClass(tempDir, clazz);
//		var transformedClass = transform(sut, SomeClass.class);
//		firstConstructor(transformedClass).newInstance("", "", "", "", false);
	}

	private static Constructor<?> firstConstructor(Class<?> clazz) {
		return Stream.of(clazz.getConstructors()).findFirst().get();
	}

	private static Class<?> transform(Plugin plugin, Class<?> clazz) throws ClassNotFoundException {
		var builder = new ByteBuddy().redefine(clazz);
		var transformed = plugin.apply(builder, TypeDescription.ForLoadedType.of(clazz), null).make();
		var classname = clazz.getName();
		var classLoader = new ByteArrayClassLoader(getSystemClassLoader(),
				singletonMap(classname, transformed.getBytes()));
		return classLoader.loadClass(classname);
	}

	private String decompileClass(Path destination, Class<?> clazz) throws IOException {
		Map<String, Object> options = Map.of( //
				IFernflowerPreferences.REMOVE_BRIDGE, "true", //
				IFernflowerPreferences.DECOMPILE_GENERIC_SIGNATURES, "true" //
		);
		ConsoleDecompiler decompiler = new ConsoleDecompiler(destination.toFile(), options);
		decompiler.addSpace(destination.toFile(), true);
		decompiler.decompileContext();

		Path decompiledJavaFile = destination.resolve(clazz.getSimpleName() + ".java");
		String decompiledSource = Files.readString(decompiledJavaFile);
		Files.delete(decompiledJavaFile);
		return decompiledSource;
	}

}
