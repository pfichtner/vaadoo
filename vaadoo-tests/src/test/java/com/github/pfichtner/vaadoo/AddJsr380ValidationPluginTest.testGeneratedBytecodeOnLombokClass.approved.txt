package com.example;

import com.example.SomeLombokClass.SomeLombokClassBuilder;
import com.google.common.base.Preconditions;
import jakarta.validation.constraints.AssertFalse;
import jakarta.validation.constraints.AssertTrue;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Null;
import java.util.Collection;
import java.util.Map;
import lombok.Generated;

public class SomeLombokClass {
   @Null
   private final Object someNullObject;
   @NotNull
   private final Object someObject;
   private final String valueWithoutAnnotation;
   @NotEmpty
   private final CharSequence someNotEmptyCharSequence;
   @NotEmpty
   private final String someNotEmptyString;
   @NotEmpty
   private final Collection someNotEmptyCollection;
   @NotEmpty
   private final Map someNotEmptyMap;
   @NotEmpty
   private final Integer[] someNotEmptyArray;
   @NotBlank
   private final CharSequence someNonBlankValue;
   @NotBlank(
      message = "my custom message"
   )
   private final String someNonBlankValueWithCustomMessage;
   @AssertTrue
   private final boolean someTrueValue;
   @AssertFalse
   private final boolean someFalseValue;
   @AssertTrue
   private final Boolean someTrueValueWrapper;
   @AssertFalse
   private final Boolean someFalseValueWrapper;
   @Min(42L)
   private final int someIntPrimitiveValueThatIsMinimal42;
   @Min(42L)
   private final Long someLongWrapperValueThatIsMinimal42;
   @NotNull
   @Min(42L)
   @Max(42L)
   private final Short someShortWrapperValueThatIsNotNullAndMinimal42;

   public static void main(String[] args) {
      System.out.println(builder().someObject("isNotNull").someNonBlankValue("").build());
   }

   @Generated
   SomeLombokClass(@Null Object someNullObject, @NotNull Object someObject, String valueWithoutAnnotation, @NotEmpty CharSequence someNotEmptyCharSequence, @NotEmpty String someNotEmptyString, @NotEmpty Collection someNotEmptyCollection, @NotEmpty Map someNotEmptyMap, Integer[] someNotEmptyArray, @NotBlank CharSequence someNonBlankValue, @NotBlank(message = "my custom message") String someNonBlankValueWithCustomMessage, @AssertTrue boolean someTrueValue, @AssertFalse boolean someFalseValue, @AssertTrue Boolean someTrueValueWrapper, @AssertFalse Boolean someFalseValueWrapper, @Min(42L) int someIntPrimitiveValueThatIsMinimal42, @Min(42L) Long someLongWrapperValueThatIsMinimal42, @NotNull @Min(42L) Short someShortWrapperValueThatIsNotNullAndMinimal42) {
      validate(someNullObject, someObject, valueWithoutAnnotation, someNotEmptyCharSequence, someNotEmptyString, someNotEmptyCollection, someNotEmptyMap, someNotEmptyArray, someNonBlankValue, someNonBlankValueWithCustomMessage, someTrueValue, someFalseValue, someTrueValueWrapper, someFalseValueWrapper, someIntPrimitiveValueThatIsMinimal42, someLongWrapperValueThatIsMinimal42, someShortWrapperValueThatIsNotNullAndMinimal42);
      this.someNullObject = someNullObject;
      this.someObject = someObject;
      this.valueWithoutAnnotation = valueWithoutAnnotation;
      this.someNotEmptyCharSequence = someNotEmptyCharSequence;
      this.someNotEmptyString = someNotEmptyString;
      this.someNotEmptyCollection = someNotEmptyCollection;
      this.someNotEmptyMap = someNotEmptyMap;
      this.someNotEmptyArray = someNotEmptyArray;
      this.someNonBlankValue = someNonBlankValue;
      this.someNonBlankValueWithCustomMessage = someNonBlankValueWithCustomMessage;
      this.someTrueValue = someTrueValue;
      this.someFalseValue = someFalseValue;
      this.someTrueValueWrapper = someTrueValueWrapper;
      this.someFalseValueWrapper = someFalseValueWrapper;
      this.someIntPrimitiveValueThatIsMinimal42 = someIntPrimitiveValueThatIsMinimal42;
      this.someLongWrapperValueThatIsMinimal42 = someLongWrapperValueThatIsMinimal42;
      this.someShortWrapperValueThatIsNotNullAndMinimal42 = someShortWrapperValueThatIsNotNullAndMinimal42;
   }

   @Generated
   public static SomeLombokClassBuilder builder() {
      return new SomeLombokClassBuilder();
   }

   private static void validate(Object var0, Object var1, String var2, CharSequence var3, String var4, Collection var5, Map var6, Integer[] var7, CharSequence var8, String var9, boolean var10, boolean var11, Boolean var12, Boolean var13, int var14, Long var15, Short var16) {
      Preconditions.checkArgument(var0 == null, "someNullObject must be null");
      Preconditions.checkNotNull(var1, "someObject must not be null");
      Preconditions.checkArgument(((CharSequence)Preconditions.checkNotNull(var3, "someNotEmptyCharSequence must not be empty")).length() > 0, "someNotEmptyCharSequence must not be empty");
      Preconditions.checkArgument(((CharSequence)Preconditions.checkNotNull(var4, "someNotEmptyString must not be empty")).length() > 0, "someNotEmptyString must not be empty");
      Preconditions.checkArgument(((Collection)Preconditions.checkNotNull(var5, "someNotEmptyCollection must not be empty")).size() > 0, "someNotEmptyCollection must not be empty");
      Preconditions.checkArgument(((Map)Preconditions.checkNotNull(var6, "someNotEmptyMap must not be empty")).size() > 0, "someNotEmptyMap must not be empty");
      Preconditions.checkArgument(((Object[])Preconditions.checkNotNull(var7, "someNotEmptyArray must not be empty")).length > 0, "someNotEmptyArray must not be empty");
      Preconditions.checkArgument(((CharSequence)Preconditions.checkNotNull(var8, "someNonBlankValue must not be blank")).toString().trim().length() > 0, "someNonBlankValue must not be blank");
      Preconditions.checkArgument(((CharSequence)Preconditions.checkNotNull(var9, "my custom message")).toString().trim().length() > 0, "my custom message");
      Preconditions.checkArgument(var10, "someTrueValue must be true");
      Preconditions.checkArgument(!var11, "someFalseValue must be false");
      Preconditions.checkArgument(var12 == null || var12.booleanValue(), "someTrueValueWrapper must be true");
      Preconditions.checkArgument(var13 == null || !var13.booleanValue(), "someFalseValueWrapper must be false");
      Preconditions.checkArgument((long)var14 >= 42L, "someIntPrimitiveValueThatIsMinimal42 must be greater than or equal to 42");
      Preconditions.checkArgument(var15 == null || var15.longValue() >= 42L, "someLongWrapperValueThatIsMinimal42 must be greater than or equal to 42");
      Preconditions.checkNotNull(var16, "someShortWrapperValueThatIsNotNullAndMinimal42 must not be null");
      Preconditions.checkArgument(var16 == null || (long)var16.shortValue() >= 42L, "someShortWrapperValueThatIsNotNullAndMinimal42 must be greater than or equal to 42");
   }
}
